docs/developer/adr/0002-remove-special-handling-of-optional-and-union.md

```diff
@@ -0,0 +1,92 @@
+# ADR 0002: Remove special handling of Optional and Union
+
+- Status: proposed
+- Deciders:
+- Date: 2024-04-11
+
+## Context
+
+### General
+
+Sciline builds a data dependency graph based on type hints of callables.
+Some callables may have optional inputs, which are commonly represented by `Optional[T]` in the type hint, for some type `T`.
+Therefore, in [#50](https://github.com/scipp/sciline/pull/50) we have added special handling for `Optional` and [#89](https://github.com/scipp/sciline/pull/89) extended this for `Union`.
+In the case of `Optional`, they way this works is that `sciline.Pipeline` prunes branches at the node where the optional input used, if any ancestor node has unsatisfied dependencies.
+Instead, an implicit `None` provider is added.
+This has a series of problems, which we exemplify for the case of `Optional`.
+
+1. Default values (which are currently ignored by Sciline) are overridden by the implicit `None` provider.
+   In other words, Sciline assumes that the default value of the optional input is `None`.
+2. Entire branches a pruned, which can hide bugs.
+   If the users added providers for the optional input, they will not be used if any of them has unintentionally unsatisfied dependencies.
+3. The special mechanism prevents the (in principle very valid) use of any providers that return an `Optional` or `Union` type.
+4. Optional inputs cannot be set to `None` *explicitly*.
+
+In summary, the special handling of `Optional` and `Union` is too implicit and causes more problems than it solves.
+There are a couple more aspects to consider.
+
+### Readability of user code
+
+Handling `Optional` explicitly would make user code more readable.
+Consider the following example:
+
+```python
+pipeline[MyParam] = 1.2
+```
+
+In the current implementation this gives no indication to the user that `MyParam` is not a required input.
+Furthermore, if the line is removed, the user may not realize that `MyParam` is available as an optional input.
+With the proposed change, the user can make this explicit:
+
+```python
+pipeline[Optional[MyParam]] = 1.2
+```
+
+Above it is clear that `MyParam` is optional, and it can be set to `None` explicitly:
+
+```python
+pipeline[Optional[MyParam]] = None
+```
+
+### Code complexity and maintainability
+
+The special handling of `Optional` and `Union` is a significant source of complexity in the code, requiring a significant amount of unit testing.
+
+### Conceptual clarity
+
+The current redesign of Sciline highlighted that the current implementation is conceptually flawed.
+It makes it tricky to represent the internals of `sciline.Pipeline` as a simple data dependency graph.
+The special handling of `Optional` and `Union` seems to require pervasive changes to the code, which is a sign that it is not a good fit for the design.
+
+### Counter arguments
```

Comment:
Not sure I follow. `load_sample` would not need to change with the proposal?

Reply:
Did I misunderstand it? Are you saying you want to keep support for `Union` but remove the branch-pruning behaviour of `Optional`?

Reply:
The user can still use `Union` and `Optional`, but Sciline would not longer handle them in any special manner. A provider could then return `Union[int, float]`, and another provider could depend on this.

What would be removed is the logic to hook up `int` or `float` to `Union[int, float]`.

Reply:
And that is exactly why I said it is bad for the example. It would require specifying the parameter as
```python
params = {
  Union[FilePath, NeXusFile, NeXusGroup]: Path("...")
}
```
which is unwieldy. The best way to improve this I could think of is
```python
FileSpec = NewType("FileSpec", Union[FilePath, NeXusFile, NeXusGroup])
params = {
  FileSpec: Path("...")
}
```
but this is odd because it requires two explicit types (key and value) but those are not the same.